django_cloud_tasks/signals.py

```diff
@@ -0,0 +1,88 @@
+# pylint: disable=no-member
+from django.core.exceptions import ValidationError
+from django.db.models.signals import pre_save
+from django.dispatch import receiver
+from django.db.models import Model
+
+from django_cloud_tasks import models, tasks
+
+
+@receiver(pre_save, sender=models.Routine)
+def ensure_valid_task_name(sender, instance, **kwargs):
+    # A exception is raised when the task is not found
+    instance.task  # pylint: disable=pointless-statement
+
+
+def _is_status_changing(instance: Model) -> bool:
+    if not instance.pk:
+        return False
+
+    current_routine = models.Routine.objects.get(pk=instance.pk)
+
+    return current_routine.status != instance.status
+
+
+@receiver(pre_save, sender=models.Routine)
+def enqueue_next_routines(sender, instance, **kwargs):
+    if not _is_status_changing(instance=instance):
+        return
+
+    if instance.status == models.Routine.Statuses.COMPLETED:
+        for vertice in instance.next_routine_vertices.all():
+            vertice.next_routine.enqueue()
+
+
+@receiver(pre_save, sender=models.Routine)
+def enqueue_previously_routines(sender, instance, **kwargs):
+    if not _is_status_changing(instance=instance):
+        return
+
+    if instance.status == models.Routine.Statuses.REVERTED:
+        for vertice in instance.required_routine_vertices.all():
+            vertice.routine.revert()
+
+
+@receiver(pre_save, sender=models.Routine)
+def enqueue_routine_scheduled(sender, instance, **kwargs):
+    if not _is_status_changing(instance=instance):
+        return
+
+    if instance.status == models.Routine.Statuses.SCHEDULED:
+        tasks.PipelineRoutineTask().delay(routine_id=instance.pk)
+
+
+@receiver(pre_save, sender=models.Routine)
+def enqueue_revert_task(sender, instance, **kwargs):
+    if not _is_status_changing(instance=instance):
+        return
+
+    if instance.status == models.Routine.Statuses.REVERTING:
+        meta = {"routine_id": instance.pk}
+        instance.task().revert(data=instance.output, _meta=meta)
+
+
+@receiver(pre_save, sender=models.Routine)
+def ensure_status_machine(sender, instance, **kwargs):
+    if not instance.pk and instance.status != models.Routine.Statuses.PENDING:
+        raise ValidationError(f"The initial routine's status must be 'pending' not '{instance.status}'")
+
+    if not _is_status_changing(instance=instance):
```

Comment:
This is the 5th pre_save signal that fetches the existing status. It means every save will trigger at least 5 db reads. And you are now vulnerable to the order they execute (which the order Django saves each method into the callback array).

My suggestion is to move it all to one handler that is smart enough to detect which vertex of the status state-machine the Routine is on, and them take action.